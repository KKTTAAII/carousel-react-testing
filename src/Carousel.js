import React, { useState } from "react";
import "./Carousel.css";
import image1 from "./assets/image1.jpg";
import image2 from "./assets/image2.jpg";
import image3 from "./assets/image3.jpg";
import Card from "./Card";

const Carousel = ({ cardData, title }) => {
  const [cardIdx, setCardIdx] = useState(0);
  const card = cardData[cardIdx];
  const total = cardData.length;
  const goForward = () => setCardIdx(cardIdx + 1);
  const goBackward = () => setCardIdx(cardIdx - 1);
  const leftClass = `fas fa-chevron-circle-left fa-2x`
  const rightClass = `fas fa-chevron-circle-right fa-2x`
  const {caption, src} = card;

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">
        <i
          className={
            cardIdx === 0
              ? leftClass + " hide"
              : leftClass
          }
          onClick={goBackward}
          data-testid="left-arrow"
        />
        <Card
          caption={caption}
          src={src}
          currNum={cardIdx + 1}
          totalNum={total}
        />
        <i
          className={
            cardIdx === total - 1
              ? rightClass + " hide"
              : rightClass
          }
          onClick={goForward}
          data-testid="right-arrow"
        />
      </div>
    </div>
  );
};

Carousel.defaultProps = {
  cardData: [
    {
      src: image1,
      caption: "Photo by Richard Pasquarella on Unsplash",
    },
    {
      src: image2,
      caption: "Photo by Pratik Patel on Unsplash",
    },
    {
      src: image3,
      caption: "Photo by Josh Post on Unsplash",
    },
  ],
  title: "Shells from far away beaches.",
};

export default Carousel;
