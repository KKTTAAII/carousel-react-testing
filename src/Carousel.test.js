import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";

it("works when you click on the right arrow", () => {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  // expect the first image to show, but not the second
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = queryByTestId("right-arrow");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).not.toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).toBeInTheDocument();
});

it("display the previous picture when using left arrow", () => {
  const { queryByTestId, queryByAltText } = render(<Carousel />);

  const rightArrow = queryByTestId("right-arrow");
  const leftArrow = queryByTestId("left-arrow");
  fireEvent.click(rightArrow);
  fireEvent.click(leftArrow);

  //expect the current image to be the previous image 
  expect(queryByAltText("Photo by Richard Pasquarella on Unsplash")).toBeInTheDocument();
  expect(queryByAltText("Photo by Pratik Patel on Unsplash")).not.toBeInTheDocument();
});

it("should hide arrow when on the first or last picture", () => {
  const { queryByTestId } = render(<Carousel />);
  const rightArrow = queryByTestId("right-arrow");
  const leftArrow = queryByTestId("left-arrow");

  //the left arrow should hide when on the first picture
  expect(leftArrow).toHaveClass("fas fa-chevron-circle-left fa-2x hide");

  fireEvent.click(rightArrow);
  fireEvent.click(rightArrow);
  
  //the right arrow should hide when on the last picture
  expect(rightArrow).toHaveClass("fas fa-chevron-circle-right fa-2x hide");
});

it("renders Carousel component successfully", () => {
  render(<Carousel />)
});

it("matches snapshot", () => {
  const {asFragment} = render(<Carousel />);
  expect(asFragment()).toMatchSnapshot();
});