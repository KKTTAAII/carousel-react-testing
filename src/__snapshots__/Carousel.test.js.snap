// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      Shells from far away beaches.
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="fas fa-chevron-circle-left fa-2x hide"
        data-testid="left-arrow"
      />
      <div
        class="Card"
      >
        <span
          class="Card-title"
        >
          Photo by Richard Pasquarella on Unsplash
        </span>
        <img
          alt="Photo by Richard Pasquarella on Unsplash"
          class="Card-image"
          src="image1.jpg"
        />
        <span
          class="Card-small"
        >
          Image 1 of 3.
        </span>
      </div>
      <i
        class="fas fa-chevron-circle-right fa-2x"
        data-testid="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;
